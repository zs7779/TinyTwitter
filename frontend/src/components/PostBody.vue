<template>
    <div :class="[hovering ? 'hover':'', 'p-3']" @mouseover='hovering=true' @mouseout='hovering=false'>
        <router-link :to='{name: "post", params: {username: post.author.username, postID: post.id}}' tag='div' :class="[verbose ? 'card-view':'', 'card-text']">
            <h6 class="card-title mb-1">
                <router-link :to="{ name: 'user', params: {username: post.author.username} }">{{ post.author.username }}</router-link>
                <span class="small text-muted">at {{ post.create_time }}:</span>
            </h6>
            <div v-if='post.is_comment'>
                <span class="small text-muted">Reply to</span>
                <span class="small text-muted" v-for='m in mentions' :key="m.id">
                    <router-link :to="{ name: 'user', params: {username: m.username} }">@{{ m.username }}</router-link>
                    <span> </span>
                </span>
            </div>
            <div>
                <p>{{ post.text }}</p>
            </div>
            <slot></slot>
        </router-link>
        <div v-if=buttons class="card-footer bg-transparent p-0">
            <div class="d-flex justify-content-around">
                <button type="button" v-b-modal.new-post class="btn btn-transparent px-1 py-0" title="Comment" @click="$emit('action-comment')">
                    <b-icon icon="chat" :class="[post.commented ? 'card-button-comment' : 'card-button']"></b-icon> {{ post.comment_count }}
                </button>
                <button type="button" v-b-modal.new-post class="btn btn-transparent px-1 py-0" title="Repost" @click="$emit('action-repost')">
                    <b-icon icon="arrow-repeat" :class="[post.reposted ? 'card-button-repost' : 'card-button']"></b-icon> {{ post.repost_count }}
                </button>
                <button type="button" class="btn btn-transparent px-1 py-0" title="Like" @click="$emit('action-like')">
                    <b-icon icon="heart" :class="[post.liked ? 'card-button-like' : 'card-button']"></b-icon> {{ post.like_count }}
                </button>
                <b-dropdown id="dropdown-dropup" dropup variant="btn btn-transparent px-1 py-0" no-caret title="Options">
                    <template v-slot:button-content><b-icon icon="chevron-up" class="card-button"></b-icon></template>
                    <b-dropdown-item v-if="post.owner" @click="$emit('action-delete')"><b-icon icon="x" class="card-button"></b-icon> Delete</b-dropdown-item>
                    <b-dropdown-item v-b-modal.new-post v-if="post.owner" @click="$emit('action-edit')"><b-icon icon="pencil" class="card-button"></b-icon> Edit</b-dropdown-item>
                    <b-dropdown-item><b-icon icon="share" class="card-button"></b-icon> Share</b-dropdown-item>
                </b-dropdown>
            </div>
        </div>
    </div>
</template>

<script>
export default{
    props: ['post', 'buttons', 'verbose'],
    data() {
        return {
            hovering: false,
        }
    },
    computed: {
        mentions() {
            var receipants = {};
            if (this.post.root_post) {
                receipants[this.post.root_post.author.id] = this.post.root_post.author.username;
            }
            if (this.post.parent) {
                receipants[this.post.parent.author.id] = this.post.parent.author.username;
            }
            if (this.post.mentions) {
                this.post.mentions.forEach(m => {
                    receipants[m.id] = m.username;
                })
            }
            return Object.entries(receipants).map(([k, v]) => {
                return { id: k, username: v, }
            });
        },
    }
}
</script>

<style scoped>
.hover {
    background-color: #f9f9f9;
}
.card-button {
    color: #909090;
}

.card-button-comment {
    color: darkturquoise;
}
.card-button-repost {
    color: limegreen;
}
.card-button-like {
    color: red;
}
</style>