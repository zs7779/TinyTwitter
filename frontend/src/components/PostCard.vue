<template>
    <div class="card p-3">
        <new-post v-if="editMode" v-bind:oldPost="post" v-bind:noPost="true" v-on:post-ok="onEdit($event)">
            <button type="button" class="btn btn-outline-secondary rounded-pill py-0" v-on:click.prevent="editMode=false">Cancel</button>
        </new-post>
        <router-link :to='{name: "post", params: {username: post.author.username, post_id: post.id}}' tag='div' class="card-text card-view" v-else>
            <h6 class="card-title mb-1">
                <router-link :to="{ name: 'user', params: {username: post.author.username} }">{{ post.author.username }}</router-link>
                <span class="small text-muted">at {{ post.create_time }} said:</span>
            </h6>
            <p>{{ post.text }}</p>
        </router-link>
        <div class="card-footer bg-transparent p-0">
            {{ post.like_count }} likes
        </div>
        <div class="card-footer bg-transparent p-0">
            <div class="d-flex justify-content-around">
                <button type="button" class="btn btn-transparent px-1 py-0" title="Comment">
                    <b-icon icon="chat" class="card-button"></b-icon>
                </button>
                <button type="button" class="btn btn-transparent px-1 py-0" title="Repost">
                    <b-icon icon="arrow-repeat" class="card-button"></b-icon>
                </button>
                <button type="button" class="btn btn-transparent px-1 py-0" title="Like" v-on:click="onLike">
                    <b-icon icon="heart" v-bind:class="[post.liked ? 'card-button-active' : 'card-button', '']"></b-icon> {{ post.like_count }}
                </button>
                <b-dropdown id="dropdown-dropup" dropup variant="btn btn-transparent px-1 py-0" no-caret title="Options">
                    <template v-slot:button-content><b-icon icon="chevron-up" class="card-button"></b-icon></template>
                    <b-dropdown-item v-if="post.owner" v-on:click="onDelete"><b-icon icon="x" class="card-button"></b-icon> Delete</b-dropdown-item>
                    <b-dropdown-item v-if="post.owner" v-on:click="editMode=true"><b-icon icon="pencil" class="card-button"></b-icon> Edit</b-dropdown-item>
                    <b-dropdown-item><b-icon icon="share" class="card-button"></b-icon> Share</b-dropdown-item>
                </b-dropdown>
            </div>
        </div>
    </div>
</template>

<script>
import NewPost from './NewPost.vue'
import { printError, getToken, viewsMixin } from './utils'


export default{
    name: "post-card",
    props: {
        post: Object,
    },
    data() {
        return {
            editMode: false,
        }
    },
    methods: {
        onLike: function(event) {
            const token = getToken();
            if (!token) {return;}
            axios.put(`/posts/${this.post.id}`, {
                like: !this.post.liked,
            }, {
                headers: {
                    'X-CSRFTOKEN': token,
                },
            }).then(response => {
                const post = {
                    ...this.post,
                    like_count: this.post.liked ? this.post.like_count - 1 : this.post.like_count + 1,
                    liked: !this.post.liked,
                }
                this.$emit("edit-ok", post);
            }, printError)
        },
        onEdit: function(postText) {
            const token = getToken();
            if (!token) {return;}
            axios.put(`/posts/${this.post.id}`, {
                postText: postText,
            }, {
                headers: {
                    'X-CSRFTOKEN': token,
                },
            }).then(response => {
                const post = {
                    ...this.post,
                    text: postText,
                }
                this.$emit("edit-ok", post);
                this.editMode = false;
            }, printError)
        },
        onDelete: function(event) {
            const token = getToken();
            if (!token) {return;}
            axios.delete(`/posts/${this.post.id}`, {
                headers: {
                    'X-CSRFTOKEN': token,
                },
            }).then(response => {
                this.$emit("delete-ok", this.post.id);
            }, printError)
        },
    },
    components: {
        NewPost
    }
}
</script>

<style scoped>
</style>