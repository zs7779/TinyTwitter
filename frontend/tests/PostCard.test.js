import { shallowMount } from '@vue/test-utils'
import PostCard from '../src/components/PostCard'
import PostBody from '../src/components/PostBody'
import mockData from './mock_data'


jest.mock('../src/components/utils', ()=>{
    return {
        getToken: () => null // null cause axios not called
    }
})



const factory = (post={}, verbose=false) => {
    return shallowMount(PostCard, {
        propsData: {
            post: {
                ...mockData.posts[0],
                ...post,
            },
            verbose
        },
    });
}


describe('PostCard render', () => {
    let wrapper;
    afterEach(() => {
        wrapper.destroy();
    });
    it('Render default', () => {
        wrapper = factory();
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Render verbose', () => {
        wrapper = factory({}, false);
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Render repost', () => {
        wrapper = factory({parent: mockData.posts[1]}, false);
        expect(wrapper.html()).toMatchSnapshot();
    });
})

describe('PostCard Events', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = wrapper = factory({});
    });
    afterEach(() => {
        wrapper.destroy();
    });
    it('Action comment', () => {
        wrapper.findComponent(PostBody).vm.$emit('action-comment');
        expect(wrapper.emitted('action-post').length).toBe(1);
    });
    it('Action repost', () => {
        wrapper.findComponent(PostBody).vm.$emit('action-repost');
        expect(wrapper.emitted('action-post').length).toBe(1);
    });
    it('Action edit', () => {
        wrapper.findComponent(PostBody).vm.$emit('action-edit');
        expect(wrapper.emitted('action-post').length).toBe(1);
    });
    it('Action like', () => {
        const spy = jest.spyOn(wrapper.vm, 'onLike')
        wrapper.findComponent(PostBody).vm.$emit('action-like');
        expect(spy).toHaveBeenCalled();
    });
    it('Action delete', () => {
        const spy = jest.spyOn(wrapper.vm, 'onDelete')
        wrapper.findComponent(PostBody).vm.$emit('action-delete');
        expect(spy).toHaveBeenCalled();
    });
})