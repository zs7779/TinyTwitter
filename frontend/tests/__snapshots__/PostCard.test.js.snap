// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard render Render default 1`] = `
"<div class=\\"card p-1 rounded-0\\">
  <post-body-stub buttons=\\"true\\" post=\\"[object Object]\\">
    <!---->
    <!---->
  </post-body-stub>
  <!---->
</div>"
`;

exports[`PostCard render Render repost 1`] = `
"<div class=\\"card p-1 rounded-0\\">
  <post-body-stub buttons=\\"true\\" post=\\"[object Object]\\">
    <post-body-stub post=\\"[object Object]\\" class=\\"card p-3\\"></post-body-stub>
    <!---->
  </post-body-stub>
  <!---->
</div>"
`;

exports[`PostCard render Render verbose 1`] = `
"<div class=\\"card p-1 rounded-0\\">
  <post-body-stub buttons=\\"true\\" post=\\"[object Object]\\">
    <!---->
    <!---->
  </post-body-stub>
  <!---->
</div>"
`;
